### ISSUE TITLE
[rfc] Allowing plugins to perform "manual" data collection

### ISSUE THREAD

**Issue by TurboTurtle** [sosreport/sos#2992]
This is something that the RH team has discussed a bit internally, and I wanted to open it up to the wider audience for comments/feedback before pressing forward on it.

For some time our design guidelines for plugin collections have been "must be an existing command or file", with only limited exceptions (e.g. the `hashes` option to the `python` plugin). While this makes for easier maintenance, it can leave something to be desired in a collection utility. A couple reasons for this have been/are:

1. Avoiding shell scripting in plugins that grok command output between piped commands
2. The current API doesn't have a way to ensure these happen within a plugin's timeout - we generally only expose `setup()`, which is the wrong place to do actual data compilation
3. The general thought that if the data being compiled is truly important, it would be better served being shipped as part of the component's packaging.

I am proposing that we open up the plugin API a bit to allow for ad-hoc style collections. This would mean expanding the API to provide a structure for these collections to happen in a safe and predictable manner. There would also need to be guardrails in place to ensure we don't violate concern 1 above (which I think is incredibly important).

For point 2 - I think this can be easily solved. We re-wire `Plugin()` to allow plugin authors to implement their own `collect()` methods, which will get called at the end of the various `collect_*()` methods that collect whatever has been defined during `setup()`. We can further provide wrappers to make it as easy as possible to contain collections within the plugin space, e.g.:

```python
class MyPlugin(Plugin, IndependentPlugin):

    def setup(self):
        [....]

    def collect(self):  # this becomes a standard entrypoint like setup() has been for years
        with self.get_plugin_file('myfile') as pf:  # creates sos_commands/$plugin/myfile and returns the open object
            content = self.exec_cmd('foobar')
            [...do something with content...]
            pf.write(content)
```

We have a few spots already in the project and/or as requests that would be served extremely well by this kind of functionality - the `hashes` option in the `python` plugin could be moved to `collect()`, #1350 could finally be addressed safely, etc...

While this doesn't directly address point 3 above, it does serve a related point. End users and support engineers need a given set of data sooner rather than later. In most cases, it is far more likely that such collections could be delivered faster via `sos` than the various component(s) that the collections could serve. Further, I'm sure many of the projects that ship these components may not be all too keen to start shipping diag scripts just for sos collection.

I'm all onboard with this kind of evolution to sos plugins personally, however I'm certainly up for hearing counter arguments. We're way to close to closing sos-4.4, so this would be a sos-4.5 thing in my mind (or is this a sufficient enough design change to bump to sos-5.0?)

Tagging for visibility and soliciting feedback: @pmoravec @jjansky1 @jcastill @ddstreet @BryanQuigley @arif-ali @bmr-cymru

**Comment by BryanQuigley** [comment]
Agreed either moving from setup - (or having setup limited) is a good step.  The python example is great - I don't see any use for a local sysadmin to want to run that...
